### PR TITLE
Fix #728 - Support case class with more 22 fields in Scala 3

### LINKED ISSUE #728: 2.10.0-RC6 macro does not support large case class (throw ClassNotFoundException: scala.Tuple25)
### Play JSON Version (2.5.x / etc)

2.10.0-RC6

### API (Scala / Java / Neither / Both)

Scala

### Operating System

all

### JDK

all

### Library Dependencies

only play-json 2.10.0-RC6. see following build.sbt

### Expected Behavior

derive Json type classes.
**play-json 2.10.0-RC5 does not have this issue if set `scalacOptions += "-Xmax-inlines:100"` .**

### Actual Behavior

throw ClassNotFoundException

#### build.sbt

```scala
libraryDependencies += "com.typesafe.play" %% "play-json" % "2.10.0-RC6"

scalaVersion := "3.1.1"
```

```scala
package foo

case class A(
  a1: Int,
  a2: Int,
  a3: Int,
  a4: Int,
  a5: Int,
  a6: Int,
  a7: Int,
  a8: Int,
  a9: Int,
  a10: Int,
  a11: Int,
  a12: Int,
  a13: Int,
  a14: Int,
  a15: Int,
  a16: Int,
  a17: Int,
  a18: Int,
  a19: Int,
  a20: Int,
  a21: Int,
  a22: Int,
  a23: Int,
  a24: Int,
  a25: Int
)

object A {
  val format = play.api.libs.json.Json.format[A]
}
```

```
[error] 32 |  val format = play.api.libs.json.Json.format[A]
[error]    |               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[error]    |Exception occurred while executing macro expansion.
[error]    |java.lang.ClassNotFoundException: scala.Tuple25
[error]    |	at java.net.URLClassLoader.findClass(URLClassLoader.java:387)
[error]    |	at java.lang.ClassLoader.loadClass(ClassLoader.java:418)
[error]    |	at java.lang.ClassLoader.loadClass(ClassLoader.java:351)
[error]    |	at java.lang.Class.forName0(Native Method)
[error]    |	at java.lang.Class.forName(Class.java:264)
[error]    |	at play.api.libs.json.QuotesHelper.withTuple(QuotesHelper.scala:163)
[error]    |	at play.api.libs.json.QuotesHelper.withTuple$(QuotesHelper.scala:18)
[error]    |	at play.api.libs.json.JsMacroImpl$$anon$3.withTuple(JsMacroImpl.scala:484)
[error]    |	at play.api.libs.json.JsMacroImpl$WritesHelper.productWrites(JsMacroImpl.scala:677)
[error]    |	at play.api.libs.json.JsMacroImpl$WritesHelper.productWrites$(JsMacroImpl.scala:546)
[error]    |	at play.api.libs.json.JsMacroImpl$$anon$3.productWrites(JsMacroImpl.scala:484)
[error]    |	at play.api.libs.json.JsMacroImpl$.writesExpr(JsMacroImpl.scala:516)
[error]    |	at play.api.libs.json.JsMacroImpl$.inline$writesExpr(JsMacroImpl.scala:472)
[error]    |	at play.api.libs.json.JsMacroImpl$.$anonfun$281(JsMacroImpl.scala:462)
[error]    |	at play.api.libs.json.JsMacroImpl$.$anonfun$adapted$35(JsMacroImpl.scala:462)
[error]    |	at dotty.tools.dotc.quoted.PickledQuotes$$anon$1.transform(PickledQuotes.scala:81)
[error]    |	at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1430)
[error]    |	at dotty.tools.dotc.quoted.PickledQuotes$$anon$1.transform(PickledQuotes.scala:106)
[error]    |	at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1410)
[error]    |	at dotty.tools.dotc.quoted.PickledQuotes$$anon$1.transform(PickledQuotes.scala:106)
[error]    |	at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1469)
[error]    |	at dotty.tools.dotc.quoted.PickledQuotes$$anon$1.transform(PickledQuotes.scala:106)
[error]    |	at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform$$anonfun$1(Trees.scala:1500)
[error]    |	at scala.collection.immutable.List.mapConserve(List.scala:472)
[error]    |	at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1500)
[error]    |	at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transformStats(Trees.scala:1498)
[error]    |	at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1410)
[error]    |	at dotty.tools.dotc.quoted.PickledQuotes$$anon$1.transform(PickledQuotes.scala:106)
[error]    |	at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1410)
[error]    |	at dotty.tools.dotc.quoted.PickledQuotes$$anon$1.transform(PickledQuotes.scala:106)
[error]    |	at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform$$anonfun$1(Trees.scala:1500)
[error]    |	at scala.collection.immutable.List.mapConserve(List.scala:472)
[error]    |	at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1500)
[error]    |	at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1396)
[error]    |	at dotty.tools.dotc.quoted.PickledQuotes$$anon$1.transform(PickledQuotes.scala:106)
[error]    |	at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1410)
[error]    |	at dotty.tools.dotc.quoted.PickledQuotes$$anon$1.transform(PickledQuotes.scala:106)
[error]    |	at dotty.tools.dotc.quoted.PickledQuotes$.spliceTerms(PickledQuotes.scala:121)
[error]    |	at dotty.tools.dotc.quoted.PickledQuotes$.unpickleTerm(PickledQuotes.scala:60)
[error]    |	at scala.quoted.runtime.impl.QuotesImpl.unpickleExpr(QuotesImpl.scala:2973)
[error]    |	at play.api.libs.json.JsMacroImpl$.configuredWrites(JsMacroImpl.scala:462)
[error]    |	at play.api.libs.json.JsMacroImpl$.inline$configuredWrites(JsMacroImpl.scala:450)
[error]    |	at play.api.libs.json.JsMacroImpl$.formatImpl$$anonfun$2(JsMacroImpl.scala:786)
[error]    |	at play.api.libs.json.JsMacroImpl$.formatImpl$$anonfun$adapted$2(JsMacroImpl.scala:786)
[error]    |	at dotty.tools.dotc.quoted.PickledQuotes$$anon$1.transform(PickledQuotes.scala:81)
[error]    |	at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1430)
[error]    |	at dotty.tools.dotc.quoted.PickledQuotes$$anon$1.transform(PickledQuotes.scala:106)
[error]    |	at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1464)
[error]    |	at dotty.tools.dotc.quoted.PickledQuotes$$anon$1.transform(PickledQuotes.scala:106)
[error]    |	at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform$$anonfun$1(Trees.scala:1500)
[error]    |	at scala.collection.immutable.List.mapConserve(List.scala:472)
[error]    |	at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1500)
[error]    |	at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transformStats(Trees.scala:1498)
[error]    |	at dotty.tools.dotc.ast.Trees$Instance$TreeMap.transform(Trees.scala:1410)
[error]    |	at dotty.tools.dotc.quoted.PickledQuotes$$anon$1.transform(PickledQuotes.scala:106)
[error]    |	at dotty.tools.dotc.quoted.PickledQuotes$.spliceTerms(PickledQuotes.scala:121)
[error]    |	at dotty.tools.dotc.quoted.PickledQuotes$.unpickleTerm(PickledQuotes.scala:60)
[error]    |	at scala.quoted.runtime.impl.QuotesImpl.unpickleExpr(QuotesImpl.scala:2973)
[error]    |	at play.api.libs.json.JsMacroImpl$.formatImpl(JsMacroImpl.scala:786)
[error]    |	at play.api.libs.json.JsMacroImpl$.format$$anonfun$1(JsMacroImpl.scala:63)
[error]    |	at play.api.libs.json.JsMacroImpl$.withSummonedConfig(JsMacroImpl.scala:79)
[error]    |	at play.api.libs.json.JsMacroImpl$.format(JsMacroImpl.scala:63)
[error]    |
[error]    | This location contains code that was inlined from A.scala:32
[error] one error found
```

### PR DESCRIPTION
**Pull request by cchantep**
Fix #728

### CODE REVIEW

File: play-json/shared/src/test/scala-3/play/api/libs/json/QuotesSpec.scala [hunk 1]
```diff
@@ -70,20 +70,26 @@ final class QuotesSpec extends AnyWordSpec with Matchers with org.scalatestplus.
...
-        ).mustEqual("scala.Tuple2[scala.Predef.String, scala.Int]/Foo(1,2)")
+        ).mustEqual("play.api.libs.json.Foo/Foo(1,2)")
       }
 
       "from generic Bar" in {
         testWithTuple(
           Bar[Double]("bar1", None, Seq(1.2D, 34.5D))
         ).mustEqual(
-          "scala.Tuple3[scala.Predef.String, scala.Option[scala.Double], scala.collection.immutable.Seq[scala.Double]]/Bar(bar1,None,List(1.2, 34.5))"
+          "play.api.libs.json.Bar[scala.Double]/Bar(bar1,None,List(1.2, 34.5))"
```
Comment (cchantep) [inline play-json/shared/src/test/scala-3/play/api/libs/json/QuotesSpec.scala:80]: Case class type is directly usable (and more specific)

File: play-json/shared/src/main/scala-3/play/api/libs/json/QuotesHelper.scala [hunk 2]
```diff
@@ -121,16 +120,32 @@ private[json] trait QuotesHelper {
...
-        )
+    val fields = decls.zipWithIndex.map { case ((sym, t), i) =>
+      debug(
+        s"// Field: ${sym.owner.owner.fullName}.${sym.name}, type = ${t.typeSymbol.fullName}, annotations = [${sym.annotations.map(_.show).mkString(", ")}]"
+      )
+
+      val fieldNme = s"_${i + 1}"
+
+      val resolve: Term => Term = {
+        val field = tupleTpeSym.declaredField(fieldNme)
```
Comment (cchantep) [inline play-json/shared/src/main/scala-3/play/api/libs/json/QuotesHelper.scala:133]: Either resolve as field or nullary method

File: play-json/shared/src/main/scala-3/play/api/libs/json/QuotesHelper.scala [hunk 3]
```diff
@@ -143,30 +158,18 @@ private[json] trait QuotesHelper {
...
-      }
-    }
-
-    val tupleTpr = unappliedTupleTpr.appliedTo(types)
-
-    tupleTpr -> {
+      tt: Type[T],
+      ut: Type[U]
+  ): Tuple2[TypeRepr, Expr[T] => (Expr[U] => Expr[R]) => Expr[R]] =
+    TypeRepr.of(using ut) -> {
```
Comment (cchantep) [inline play-json/shared/src/main/scala-3/play/api/libs/json/QuotesHelper.scala:172]: Directly use the given product type

File: play-json/shared/src/test/scala-3/play/api/libs/json/MacroScala3Spec.scala [hunk 4]
```diff
@@ -13,14 +13,47 @@ final class MacroScala3Spec
...
-        "Macros.writer[UsingAlias]".mustNot(typeCheck)
+      "no custom ProductOf" in {
+        "Json.writes[CustomNoProductOf]" mustNot typeCheck
       }
+    }
 
-      "no custom ProductOf" in {
-        "Macros.writer[CustomNoProductOf]".mustNot(typeCheck)
+    "be handled" when {
+      "is declared with more than 22 fields" in {
```
Comment (cchantep) [inline play-json/shared/src/test/scala-3/play/api/libs/json/MacroScala3Spec.scala:22]: Test failing before

File: play-json/shared/src/main/scala/play/api/libs/json/Json.scala [hunk 5]
```diff
@@ -227,7 +227,8 @@ object Json extends JsonFacade with JsMacros with JsValueMacros {
     JsValueWrapperImpl(w.writes(field))
 
   def obj(fields: (String, JsValueWrapper)*): JsObject = JsObject(fields.map(f => (f._1, unwrap(f._2))))
-  def arr(items: JsValueWrapper*): JsArray             = JsArray(items.iterator.map(unwrap).toArray[JsValue])
+
```
Comment (cchantep) [inline play-json/shared/src/main/scala/play/api/libs/json/Json.scala:230]: Minor formatting

File: play-json/shared/src/test/scala-3/play/api/libs/json/QuotesSpec.scala [hunk 6]
```diff
@@ -143,6 +161,66 @@ case class Foo(bar: String, lorem: Int)
 
 case class Bar[T](name: String, opt: Option[T], scores: Seq[Double])
 
+case class BigFat(
```
Comment (cchantep) [inline play-json/shared/src/test/scala-3/play/api/libs/json/QuotesSpec.scala:164]: Case class with 26 fields

**Comment by raul782** [comment]
@ihostage is there a reason why this PR doesn't have any comments/feedback? I've run it locally and it seems to be working fine. The tests pass so I wonder why it's stopping us from merging and do a cutover for RC7.

**Comment by octonato** [comment]
> @ihostage is there a reason why this PR doesn't have any comments/feedback? I've run it locally and it seems to be working fine. The tests pass so I wonder why it's stopping us from merging and do a cutover for RC7.

No special reason. Probably just got out of the radar. 

There is something wrong with CI, so I will close it and reopen to retrigger it, but I think it's ready for merge.

**Pull request merged**